Implement `?` macro repetition

mark-i-m opened the pull request:
See rust-lang/rfcs#2298 (with disposition merge)

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @nikomatsakis (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

kennytm commented:
Marking as blocked by #47732, #47603, and of course rust-lang/rfcs#2298. I don't think this can be easily reviewed with all those formatting change 😂

mark-i-m commented:
@kennytm I 100% agree. Thanks :)

kennytm commented:
@mark-i-m Please put the new `?` syntax behind a feature gate since this is an unstable feature. You may refer to `:vis` (search for `macro_vis_matcher`) to know how this works.

bors commented:
:umbrella: The latest upstream changes (presumably #47748) made this pull request unmergeable. Please resolve the merge conflicts.

mark-i-m commented:
@kennytm Done.

mark-i-m commented:
This doesn't have a tracking issue yet, so when that happens, I should update the feature gate...

mark-i-m commented:
Also the unstable book...

nikomatsakis commented on src/libsyntax/ext/tt/quoted.rs:
```diff
@@ -223,32 +225,90 @@ fn parse_tree<I>(tree: tokenstream::TokenTree,
+        // #1 is `?` token, but it could be a Kleene::ZeroOrOne without a separator or it could
+        // be a `?` separator followed by any Kleene operator. We need to look ahead 1 token to
+        // find out which.
+        Ok(Ok(op)) => {
```
Nit: can we add `assert_eq!(op, KleeneOp::ZeroOrOne);` here? I find it helps catch mistakes when later somebody messes with the arms accidentally....

nikomatsakis commented on src/test/compile-fail/macro-at-most-once-rep-ambig.rs:
```diff
@@ -0,0 +1,26 @@
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
```
I always appreciate a comment here explaining what this test is trying to do. Example:

---

Test that in `$(a)?+`, the `?` is interpreted as a separator, but not for `$(a)?`. This test focuses on error cases.

nikomatsakis commented on src/test/run-pass/macro-at-most-once-rep.rs:
```diff
@@ -0,0 +1,25 @@
+
+macro_rules! bar {
+    ($(a)?+) => {}
+}
```
what about `$(a)?*`? we should test that too, no?

nikomatsakis commented on src/test/compile-fail/macro-at-most-once-rep-ambig.rs:
```diff
@@ -0,0 +1,26 @@
+
+macro_rules! bar {
+    ($(a)?+) => {}
+}
```
what about `$(a)?*`? we should test that too, no?

nikomatsakis commented on src/test/run-pass/macro-at-most-once-rep.rs:
```diff
@@ -0,0 +1,25 @@
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
```
comment =)

nikomatsakis commented on src/test/run-pass/macro-at-most-once-rep.rs:
```diff
@@ -0,0 +1,25 @@
+// except according to those terms.
+
+macro_rules! foo {
+    ($(a)?) => {}
```
these tests would be better if they had things in the body that made use of `a`

nikomatsakis commented on src/test/compile-fail/issue-39388.rs:
```diff
@@ -11,7 +11,7 @@
 
 macro_rules! assign {
-    (($($a:tt)*) = ($($b:tt))*) => { //~ ERROR expected `*` or `+`
+    (($($a:tt)*) = ($($b:tt))*) => { //~ ERROR 14:22: 14:29: expected one of: `*`, `+`, or `?`
```
Please move this to a `ui` test instead of hard-coding spans and things (i.e., the `14:22`)

mark-i-m replied:
Oh, I intended to remove the line/col numbers but forgot... Should I still move to a `ui` test if I remove them?

nikomatsakis replied:
Doesn't matter so much. I believe `ui` tests are generally "the future" but this test can stay a compile-fail test.

nikomatsakis commented on src/test/run-pass/macro-at-most-once-rep.rs:
```diff
@@ -12,13 +12,19 @@ macro_rules! foo {
 }
 
+macro_rules! baz {
+    ($(a),?) => {} // comma separator is meaningless for `?`
```
wait-- so `$(a),?` is *accepted* but the `,` is just a no-op? seems like it should be an error

nikomatsakis commented on src/test/ui/feature-gate-macro_at_most_once_rep.rs:
```diff
@@ -0,0 +1,19 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Test that the MSP430 interrupt ABI cannot be used when msp430_interrupt
```
This comment is wrong =)

mark-i-m commented:
@nikomatsakis 

> wait-- so $(a),? is accepted but the , is just a no-op? seems like it should be an error

How would you feel about a lint instead? I would like to keep the separator simply for consistency with `+` and `*` even though having a separator on "at most once" is silly.

mark-i-m commented:
@nikomatsakis I think I addressed your comments. You may want to wait on a re-review until #47732, #47603 are merged, as they will make the diffs easier to read...

bors commented:
:umbrella: The latest upstream changes (presumably #47870) made this pull request unmergeable. Please resolve the merge conflicts.

mark-i-m commented:
> until #47732, #47603 are merged

They have merged now :tada: ... I will rebase shortly

nikomatsakis commented:
> How would you feel about a lint instead? I would like to keep the separator simply for consistency with + and * even though having a separator on "at most once" is silly.

Maybe we should bring this up in the RFC thread and get others' opinions. I don't have a *very* strong opinion, but it seems like an error would be better to me.

mark-i-m commented:
@nikomatsakis I just rebased this PR... The diff should be _much_ more readable now. The only thing remaining is that when there is a tracking issue, I need to update the feature gate.

nikomatsakis commented:
@mark-i-m would you rather merge this now, and update the tracking issue in a follow-up PR? Seems fine to me. r=me if so.

kennytm commented:
I'd prefer just file a tracking issue now.

mark-i-m commented:
TBH, it does feel a bit weird to have a tracking issue and merged implementation before the RFC is formally accepted, but if everyone agrees, that's fine with me 👍

nikomatsakis commented:
Tracking issue created: https://github.com/rust-lang/rust/issues/48075

@mark-i-m can you update the PR to reference it?

mark-i-m commented:
Updated

CAD97 commented:
Failure is due to added `(see tracking issue)`:

```
[00:53:32] ---- [ui] ui/feature-gate-macro_at_most_once_rep.rs stdout ----
[00:53:32] 	diff of stderr:
[00:53:32] 
[00:53:32] -	error[E0658]: Using the `?` macro Kleene operator for "at most one" repetition is unstable
[00:53:32] +	error[E0658]: Using the `?` macro Kleene operator for "at most one" repetition is unstable (see issue #48075)
```

mark-i-m commented:
Thanks! Fixed.

mark-i-m commented:
hmm... tidy is now complaining that that line is too long 🤕

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit b92e542 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis b92e542ddd41affaf6fb5d1267b8e8dfc03089a5 -->

bors commented:
:hourglass: Testing commit b92e542ddd41affaf6fb5d1267b8e8dfc03089a5 with merge e8b05f5bd4c89faa3f2b4718e6582acd05da08be...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6307)

kennytm commented:
@bors retry #48116

bors commented:
:hourglass: Testing commit b92e542ddd41affaf6fb5d1267b8e8dfc03089a5 with merge b8398d947d160ad4f26cc22da66e5fbc7030817b...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6309), [status-travis](https://travis-ci.org/rust-lang/rust/builds/340199984?utm_source=github_status&utm_medium=notification)
Approved by: nikomatsakis
Pushing b8398d947d160ad4f26cc22da66e5fbc7030817b to master...
